File: distbench-py/distbench/community.py
"""Community layer for managing peer relationships.

This module provides the Community class which manages:
- Peer relationships and neighbor sets
- Address-to-peer-ID mapping
- Peer status tracking (lifecycle states)
- Connection managers for each peer
- Cryptographic public key storage
"""

import asyncio
import logging
from dataclasses import dataclass
from enum import Enum
from typing import Generic, TypeVar

from distbench.connection import ConnectionManager
from distbench.transport.base import Address, Transport

logger = logging.getLogger(__name__)


class NodeStatus(Enum):
    """Node lifecycle states.

    A node progresses through these states during its lifetime:
    - NOT_STARTED: Initial state before any operations
    - KEY_SHARING: Exchanging cryptographic public keys with peers
    - STARTING: Synchronizing startup with other nodes
    - RUNNING: Algorithm is actively executing
    - STOPPING: Coordinating graceful shutdown
    - TERMINATED: Node has completely stopped
    """

    NOT_STARTED = "not_started"
    KEY_SHARING = "key_sharing"
    STARTING = "starting"
    RUNNING = "running"
    STOPPING = "stopping"
    TERMINATED = "terminated"


@dataclass(frozen=True)
class PeerId:
    """Unique identifier for a peer in the distributed system.

    Peer IDs are immutable and hashable for use in sets and dictionaries.
    """

    id: str

    def __hash__(self) -> int:
        return hash(self.id)

    def __str__(self) -> str:
        return self.id


T = TypeVar("T", bound=Transport)
A = TypeVar("A", bound=Address)


class Community(Generic[T, A]):
    """Manages peer relationships and communication in a distributed system.

    The Community class is responsible for:
    - Tracking which peers are neighbors (direct communication)
    - Managing connection managers for each peer
    - Mapping network addresses to peer identities
    - Tracking lifecycle status of all peers
    - Storing cryptographic public keys for peers
    """

    def __init__(
        self,
        neighbours: set[PeerId],
        peer_addresses: dict[PeerId, A],
        transport: T,
    ) -> None:
        """Initialize a community with peer information.

        Args:
            neighbours: Set of peer IDs that are direct neighbors
            peer_addresses: Mapping from peer IDs to their network addresses
            transport: Transport instance for creating connections
        """
        self.neighbours = neighbours
        self.transport = transport

        # Create connection managers for all peers
        self.connections: dict[PeerId, ConnectionManager[T, A]] = {
            peer_id: ConnectionManager(transport, addr) for peer_id, addr in peer_addresses.items()
        }

        # Reverse mapping: address -> peer ID (for identifying incoming messages)
        self.peers: dict[A, PeerId] = {addr: peer_id for peer_id, addr in peer_addresses.items()}

        # Status tracking with thread-safe access
        self._peer_status: dict[PeerId, NodeStatus] = {}
        self._status_lock = asyncio.Lock()

        # Cryptographic key storage
        self._keys: dict[PeerId, bytes] = {}
        self._keystore_lock = asyncio.Lock()

        logger.trace(
            f"Community initialized with {len(self.connections)} peers, "
            f"{len(self.neighbours)} neighbors"
        )

    async def set_status(self, peer_id: PeerId, status: NodeStatus) -> None:
        """Update the status of a peer.

        Args:
            peer_id: ID of the peer to update
            status: New status for the peer
        """
        async with self._status_lock:
            old_status = self._peer_status.get(peer_id, NodeStatus.NOT_STARTED)
            self._peer_status[peer_id] = status
            if old_status != status:
                logger.trace(f"Peer {peer_id} status: {old_status.value} -> {status.value}")

    async def get_status(self, peer_id: PeerId) -> NodeStatus:
        """Get the current status of a peer.

        Args:
            peer_id: ID of the peer to query

        Returns:
            Current status of the peer (NOT_STARTED if unknown)
        """
        async with self._status_lock:
            return self._peer_status.get(peer_id, NodeStatus.NOT_STARTED)

    async def statuses(self) -> dict[PeerId, NodeStatus]:
        """Get the current status of all known peers.

        Returns:
            Dictionary mapping peer IDs to their current status
        """
        async with self._status_lock:
            return {
                peer_id: self._peer_status.get(peer_id, NodeStatus.NOT_STARTED)
                for peer_id in self.peers.values()
            }

    def id_of(self, addr: A) -> PeerId | None:
        """Look up a peer ID by its network address.

        This is used to identify the sender of incoming messages.

        Args:
            addr: Network address to look up

        Returns:
            Peer ID associated with the address, or None if not found
        """
        return self.peers.get(addr)

    def connection(self, peer_id: PeerId) -> ConnectionManager[T, A] | None:
        """Get the connection manager for a specific peer.

        Args:
            peer_id: ID of the peer

        Returns:
            Connection manager for the peer, or None if peer is unknown
        """
        return self.connections.get(peer_id)

    def get_neighbours(self) -> dict[PeerId, ConnectionManager[T, A]]:
        """Get connection managers for all neighbor peers.

        Returns:
            Dictionary mapping neighbor peer IDs to their connection managers
        """
        return {
            peer_id: self.connections[peer_id]
            for peer_id in self.neighbours
            if peer_id in self.connections
        }

    async def add_key(self, peer_id: PeerId, public_key: bytes) -> None:
        """Store a peer's cryptographic public key.

        Args:
            peer_id: ID of the peer
            public_key: Ed25519 public key bytes (32 bytes)
        """
        async with self._keystore_lock:
            self._keys[peer_id] = public_key
            logger.trace(f"Stored public key for peer {peer_id}")

    def get_key(self, peer_id: PeerId) -> bytes | None:
        """Retrieve a peer's cryptographic public key.

        Args:
            peer_id: ID of the peer

        Returns:
            Public key bytes, or None if not found
        """
        return self._keys.get(peer_id)

    async def has_all_keys(self) -> bool:
        """Check if we have public keys for all known peers.

        Returns:
            True if all peers have provided their public keys
        """
        async with self._keystore_lock:
            return all(peer_id in self._keys for peer_id in self.peers.values())

    async def close_all(self) -> None:
        """Close all connection managers.

        This should be called during shutdown to release resources.
        """
        logger.trace("Closing all connections in community")
        for conn_manager in self.connections.values():
            try:
                await conn_manager.close()
            except Exception as e:
                logger.warning(f"Error closing connection: {e}")

File: distbench-rs/lib/core/src/community.rs
//! Community management for distributed systems.
//!
//! This module defines the [`Community`] type which represents a collection of
//! peers that communicate with each other in a distributed system.

use std::{
    collections::{HashMap, HashSet},
    sync::Arc,
};

use dashmap::DashMap;
use log::trace;
use serde::{Deserialize, Serialize};

use crate::{crypto::PublicKey, status::NodeStatus, transport};

/// Unique identifier for a peer in the distributed system.
///
/// `PeerId` is used to identify nodes throughout the system. It wraps a string
/// identifier and implements the necessary traits for use as a key in collections
/// and as a transport address.
#[derive(Hash, Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
pub struct PeerId(String);

impl PeerId {
    /// Creates a new peer ID from a string.
    ///
    /// # Arguments
    ///
    /// * `id` - The string identifier for this peer
    pub fn new(id: String) -> Self {
        Self(id)
    }

    /// Converts this peer ID to a string.
    ///
    /// # Returns
    ///
    /// A clone of the underlying string identifier.
    pub fn to_string(&self) -> String {
        self.0.clone()
    }
}

/// A community of peers that communicate with each other.
///
/// The `Community` manages:
/// - The set of neighbor peers for a given node
/// - Connection managers for communicating with each peer
/// - Address-to-PeerId mappings
/// - Status tracking for all peers
/// - Public keys for peers (if applicable)
///
/// # Type Parameters
///
/// * `T` - The transport layer implementation
pub struct Community<T: transport::Transport, CM: transport::ConnectionManager<T>> {
    neighbours: HashSet<PeerId>,
    connections: HashMap<PeerId, Arc<CM>>,
    peers: HashMap<T::Address, PeerId>,
    transport: Arc<T>,
    peer_status: DashMap<PeerId, NodeStatus>,
    keys: Arc<DashMap<PeerId, PublicKey>>,
}

impl<T: transport::Transport + 'static, CM: transport::ConnectionManager<T>> Community<T, CM> {
    /// Creates a new community.
    ///
    /// # Arguments
    ///
    /// * `neighbours` - The set of peer IDs that are considered neighbors
    /// * `peer_addresses` - A mapping from peer IDs to their network addresses
    /// * `transport` - The transport layer to use for communication
    ///
    /// # Returns
    ///
    /// A new `Community` instance configured with the provided peers and transport.
    pub fn new(
        neighbours: HashSet<PeerId>,
        peer_addresses: HashMap<PeerId, T::Address>,
        transport: Arc<T>,
    ) -> Self {
        trace!(
            "Community::new - Creating community with {} neighbours, {} total peers",
            neighbours.len(),
            peer_addresses.len()
        );

        let connections: HashMap<PeerId, Arc<CM>> = peer_addresses
            .iter()
            .map(|(id, addr)| {
                trace!(
                    "Community::new - Creating connection manager for peer {} at {}",
                    id.to_string(),
                    addr
                );
                let conn_manager: Arc<CM> = Arc::new(CM::new(transport.clone(), addr.clone()));
                (id.clone(), conn_manager)
            })
            .collect();

        let peers = peer_addresses
            .iter()
            .map(|(id, addr)| (addr.clone(), id.clone()))
            .collect();

        trace!("Community::new - Community initialized");
        Self {
            neighbours,
            connections,
            peers,
            transport,
            keys: Default::default(),
            peer_status: Default::default(),
        }
    }

    /// Sets the status of a peer.
    ///
    /// # Arguments
    ///
    /// * `id` - The ID of the peer whose status to update
    /// * `status` - The new status for the peer
    pub fn set_status(&self, id: PeerId, status: NodeStatus) {
        trace!(
            "Community::set_status - Setting peer {} status to {:?}",
            id.to_string(),
            status
        );
        self.peer_status.insert(id, status);
    }

    /// Returns the current status of all known peers.
    ///
    /// # Returns
    ///
    /// A map from peer IDs to their current status. Peers with no recorded
    /// status will have the default status.
    pub fn statuses(&self) -> HashMap<PeerId, NodeStatus> {
        self.peer_status
            .iter()
            .map(|entry| (entry.key().clone(), entry.value().clone()))
            .collect()
    }

    /// Returns an iterator over connection managers for all peers.
    ///
    /// # Returns
    ///
    /// An iterator yielding connection managers for each peer in the community.
    pub fn all_peers(&self) -> impl Iterator<Item = Arc<CM>> + '_ {
        self.connections.values().cloned()
    }

    /// Looks up the peer ID for a given address.
    ///
    /// # Arguments
    ///
    /// * `addr` - The network address to look up
    ///
    /// # Returns
    ///
    /// `Some(peer_id)` if the address is known, `None` otherwise.
    pub fn id_of(&self, addr: &T::Address) -> Option<PeerId> {
        self.peers.get(addr).cloned()
    }

    /// Returns the connection manager for a specific peer.
    ///
    /// # Arguments
    ///
    /// * `id` - The ID of the peer to get the connection for
    ///
    /// # Returns
    ///
    /// `Some(connection_manager)` if the peer is known, `None` otherwise.
    pub fn connection(&self, id: &PeerId) -> Option<Arc<CM>> {
        self.connections.get(id).cloned()
    }

    /// Check if we have all the public keys for all peers.
    ///
    /// # Returns
    ///
    /// `true` if we have all the public keys for all peers, `false` otherwise.
    pub fn keystore_full(&self) -> bool {
        self.peers.values().all(|id| self.keys.contains_key(id))
    }

    /// Checks if a peer is a neighbour.
    ///
    /// # Arguments
    ///
    /// * `id` - The ID of the peer to check
    ///
    /// # Returns
    ///
    /// `true` if the peer is a neighbour, `false` otherwise.
    pub fn is_neighbour(&self, id: &PeerId) -> bool {
        self.neighbours.contains(id)
    }

    /// Returns connection managers for all neighbor peers.
    ///
    /// # Returns
    ///
    /// A map from neighbor peer IDs to their connection managers.
    pub fn neighbours(&self) -> HashMap<PeerId, Arc<CM>> {
        self.neighbours
            .iter()
            .filter_map(|id| self.connections.get(id).map(|cm| (id.clone(), cm.clone())))
            .collect()
    }

    /// Returns the transport layer used by this community.
    ///
    /// # Returns
    ///
    /// An `Arc` pointing to the transport instance.
    pub fn transport(&self) -> Arc<T> {
        self.transport.clone()
    }

    /// Sets the public key for a peer.
    ///
    /// # Arguments
    ///
    /// * `id` - The ID of the peer whose public key to set
    /// * `key` - The public key to set for the peer
    pub fn set_pubkey(&self, id: PeerId, key: PublicKey) {
        trace!(
            "Community::set_pubkey - Setting public key for peer {}",
            id.to_string()
        );
        self.keys.insert(id, key);
        trace!(
            "Community::set_pubkey - Keystore now has {}/{} keys",
            self.keys.len(),
            self.peers.len()
        );
    }

    pub fn keystore(&self) -> KeyStore {
        KeyStore {
            keys: self.keys.clone(),
        }
    }

    /// Returns the number of nodes in the community.
    pub fn size(&self) -> usize {
        self.peers.len()
    }
}

#[derive(Debug)]
pub struct KeyStore {
    keys: Arc<DashMap<PeerId, PublicKey>>,
}

impl KeyStore {
    pub fn get(
        &self,
        id: &PeerId,
    ) -> std::option::Option<dashmap::mapref::one::Ref<'_, PeerId, PublicKey>> {
        self.keys.get(id)
    }
}
